### PR TITLE
Adds Kernel.then/2

### PR DESCRIPTION
**Pull request by Eiji7**
Related issue: #10602

### CODE REVIEW

File: lib/elixir/lib/kernel.ex [hunk 1]
```diff
@@ -2455,6 +2455,26 @@ defmodule Kernel do
...
+
+  This function makes such code easy to read.
+
+  ### Examples
+
+      iex> 1 |> then(fn x -> x * 2 end)
+      2
+  """
+  @doc since: "1.12.0"
+  @spec then(x, (x -> result)) :: result when result: any, x: var
```
Comment (josevalim) [inline lib/elixir/lib/kernel.ex:2475]:
```suggestion
  @spec then(x, (x -> result)) :: result when result: var, x: var
```

File: lib/elixir/lib/kernel.ex [hunk 2]
```diff
@@ -2455,6 +2455,26 @@ defmodule Kernel do
...
   end
 
+  @doc """
+  Pipes a value into the function.
+
+  Using pipe `|>` operator you cannot use anonymous functions directly.
+  To make them work you need to wrap them into parentheses
+  and call such function like:
+
+      iex> 1 |> (fn x -> x * 2 end).()
```
Comment (wojtekmach) [inline lib/elixir/lib/kernel.ex:2465]: this approach was often discouraged (in favour of e.g. extracting private function) and with the addition of then/1 it's unnecessary. Thus, I wouldn't even talk about it. Might be worth to talk about then/1 in the pipe docs though.

Comment (Eiji7) [reply lib/elixir/lib/kernel.ex:2465]: :ok_hand: no problem

**Comment by Eiji7** [comment]
@wojtekmach Let me know if documentation looks good for you now.

File: lib/elixir/lib/kernel.ex [hunk 3]
```diff
@@ -2455,6 +2455,21 @@ defmodule Kernel do
     end
   end
 
+  @doc """
+  Pipes a value into the function.
+
+  Using pipe `|>` operator you cannot use anonymous functions directly.
+  This function makes such code easy to read.
```
Comment (josevalim) [inline lib/elixir/lib/kernel.ex:2462]:
```suggestion
  Pipes `value` into the given `fun`.

  In other words, it invokes `fun` with `value` as argument.
  This is most commonly used in pipelines, allowing you
  to pipe a value to a function outside of its first argument.
```

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Pull request merged**